common/http.hpp

```diff
@@ -0,0 +1,220 @@
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_COMMON_HTTP_HPP
+#define MENDER_COMMON_HTTP_HPP
+
+#include <common/error.hpp>
+#include <common/events.hpp>
+#include <common/expected.hpp>
+#include <common/io.hpp>
+#include <common/log.hpp>
+
+#ifdef MENDER_EVENTS_USE_BOOST
+#include <boost/asio.hpp>
+#include <boost/beast.hpp>
+#endif // MENDER_EVENTS_USE_BOOST
+
+#include <functional>
+#include <unordered_map>
+
+namespace mender {
+namespace http {
+
+using namespace std;
+
+#ifdef MENDER_EVENTS_USE_BOOST
+namespace asio = boost::asio;
+namespace beast = boost::beast;
+namespace http = beast::http;
+#endif // MENDER_EVENTS_USE_BOOST
+
+namespace error = mender::common::error;
+namespace events = mender::common::events;
+namespace expected = mender::common::expected;
+namespace io = mender::common::io;
+namespace log = mender::common::log;
+
+class HttpErrorCategoryClass : public std::error_category {
+public:
+	const char *name() const noexcept override;
+	string message(int code) const override;
+};
+extern const HttpErrorCategoryClass HttpErrorCategory;
+
+enum ErrorCode {
+	NoError = 0,
+	NoSuchHeaderError,
+	InvalidUrlError,
+	BodyMissingError,
+};
+
+error::Error MakeError(ErrorCode code, const string &msg);
+
+enum Verb {
+	GET,
+	POST,
+	PUT,
+	PATCH,
+	CONNECT,
+};
+
+string VerbToString(Verb verb);
+
+class HttpTransaction {
+public:
+	expected::ExpectedString GetHeader(const string &name) const;
+
+protected:
+	unordered_map<string, string> headers_;
+
+	friend class HttpSession;
+};
+
+using BodyGenerator = function<io::ExpectedReaderPtr()>;
+
+class HttpRequest : public HttpTransaction {
+public:
+	HttpRequest(Verb method);
+
+	error::Error SetAddress(const string &address);
+	void SetHeader(const string &name, const string &value);
+
+	// Set to a function which will generate the body. Make sure that the Content-Length set in
+	// the headers matches the length of the body. Using a generator instead of a direct reader
+	// is needed in case of redirects.
+	void SetBodyGenerator(BodyGenerator body_gen);
+
+private:
+	bool ready_ {false};
+
+	// Original address.
+	string orig_address_;
+
+	// Broken down address.
+	string protocol_;
+	string host_;
+	int port_;
+	string path_;
+
+	Verb method_;
+
+	BodyGenerator body_gen_;
+	io::ReaderPtr body_reader_;
+
+	friend class HttpSession;
+};
+
+enum StatusCode {
+	// Not a complete enum, we define only the ones we use.
+
+	StatusOK = 200,
+	StatusNotFound = 404,
+};
```

Comment:
These are not used anywhere. And all `2xx` codes are OK codes, `201` is used quite often in REST APIs.

Reply:
True, I added them preemptively. I believe the codes we will need are 200, 204 and 404. We probably won't need 201 because only the management API creates resources, the devices cannot.